src/contexts/quiz-mode/infra/llms/OpenAIQuizProvider.ts

```diff
@@ -1,69 +1,49 @@
-import { zodToJsonSchema } from 'zod-to-json-schema';
+import { Message, Role, StructuredLLMProvider } from '@/shared/application/ports/out/llm-provider';
 import { QuizProvider } from '../../application/ports/out/quiz-provider';
 import { QuizQuestion } from '../../entities/quiz-question';
-import OpenAI from 'openai';
-import { z, ZodSchema } from 'zod';
 
 export class OpenAIQuizProvider implements QuizProvider {
-    client: OpenAI;
-    
-    constructor(private readonly apiKey: string) {
-        this.client = new OpenAI({ apiKey: this.apiKey });
+    constructor(private readonly structuredLLMProvider: StructuredLLMProvider) {}
+ 
+    async generateQuestions<T extends QuizQuestion>(content: string, schema: Record<string, unknown>, numOfQuestions = 2): Promise<T[]> {
+        const arraySchema = this.wrapSchemaInArray(schema, numOfQuestions);
+
+        const result = await this.structuredLLMProvider.functionCalling(
```

Comment:
can we specify the type in template here <> so that we can avoid type reinterpret `as` in the return statement?